<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [152]:
import pandas as pd
import psycopg2

In [153]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
connection = psycopg2.connect(
dbname = 'project_sql',
user = 'skillfactory',
password = 'cCkxxLVrDE8EbvjueeMedPKt',
host = '84.201.134.129',
port = 5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
# текст запроса
query_3_1 = f'''
select count (*) 
from vacancies
'''

In [4]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
# текст 
query_3_2 = f'''
select count (*) 
from employers
'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [7]:
# текст запроса
query_3_3 = f'''
select count (*) 
from areas
'''

In [8]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [9]:
# текст запроса
query_3_4 = f'''
select count (*) 
from industries
'''

In [10]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

## выводы по предварительному анализу данных
Судя по первым четырём запросам это довольно живой рынок труда, где в среднем на каждого работодателя приходится 2,09 активных вакансий. Вакансии эти также распределены по 1362 регионам (какого количества стран пока непонятно) и 294 сферам деятельности, что говорит нам об универсальности спроса на специалистов этой профессии

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [11]:
# текст запроса
query_4_1 = f'''
select a.name area, count(distinct v.id) cnt
from areas a
join vacancies v on v.area_id = a.id
group by 1
order by 2 desc
limit 10
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [13]:
# текст запроса
query_4_2 = f'''
select count (v.id)
from vacancies v
where v.salary_from is not null or v.salary_to is not null
'''

In [14]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [15]:
# текст запроса
query_4_3 = f'''
select avg(v.salary_from), avg(v.salary_to)
from vacancies v
'''

In [16]:
# результат запроса
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\1988040164.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [17]:
# текст запроса
query_4_4 = f'''
select v.schedule schedule, v.employment, count(v.id)
   from vacancies v
   group by schedule, employment
 order by 3 DESC
'''


In [18]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [19]:
# текст запроса
query_4_5 = f'''
SELECT DISTINCT v.experience, count(*)
from vacancies v
group by 1
order by 2
'''

In [20]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Выводы по юниту 4:
1) Большинство вакансий размещали в крупных городах РФ и СНГ (Москва, СПБ, Минск, Новосибирск, Алматы и пр. )
2) Подавляющее большинство вакансий расчитаны на полный график в том или ином виде
3) Большинтсво компаний ищет сотрудников с опытом работы

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [21]:
# текст запроса
query_5_1 = f'''
(select e.name, count (v.id)
from employers e
join vacancies v on v.employer_id = e.id
group by 1
order by 2 desc
limit 1)
union
(select e.name, count (v.id)
from employers e
join vacancies v on v.employer_id = e.id
group by 1
order by 2 desc
offset 4
limit 1)
'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0  Газпром нефть    331
1         Яндекс   1933

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [23]:
# текст запроса
query_5_2 = f'''
select a.name area, count(distinct v.id) vacancy_count, count(distinct e.id) comp_count
from areas a
left join vacancies v on a.id = v.area_id
left join employers e on a.id = e.area
where v.area_id is null
group by a.name
order by  3 desc
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  vacancy_count  comp_count
0                            Россия              0         410
1                         Казахстан              0         207
2                Московская область              0          75
3                Краснодарский край              0          19
4                          Беларусь              0          18
..                              ...            ...         ...
588                        Кувандык              0           0
589                        Куйбышев              0           0
590  Куйбышево (Ростовская область)              0           0
591                        Култаево              0           0
592                        Кульсары              0           0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [25]:
# текст запроса
query_5_3 = f'''
select distinct e.id, count(distinct v.area_id)
from employers e
join vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


id  count
0         1740    181
1         2748    152
2      5724811    116
3      5130287     88
4      3682876     71
...        ...    ...
14901  8901903      1
14902  8903436      1
14903  8903831      1
14904  8904282      1
14905  8906189      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [27]:
# текст запроса
query_5_4 = f'''
select count(e.id)
from employers e
left join employers_industries ei on e.id = ei.employer_id
where industry_id is null
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [29]:
# текст запроса
query_5_5 = f'''
select e.name
from employers e
left join employers_industries ei on e.id = ei.employer_id
group by 1
having count(ei.industry_id) = 4
order by 1
offset 2
limit 1
'''

In [30]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [31]:
# текст запроса
query_5_6 = f'''
select count(distinct e.id)
from employers_industries ei
join employers e on e.id = ei.employer_id
join industries i on i.id = ei.industry_id
where i.name = 'Разработка программного обеспечения'
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [33]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
table = page.find('table', class_='standard')
tbody = table.find("tbody") # в таблице выбираем тело таблицы
rows = tbody.find_all("tr") #вычленяем ряды таблицы
million_cities = list() #создаём список, куда будем добавлять названия городов
for i in rows: #в каждом ряде находим тег "а", преобразуем его в список
    j = list(i.find('a')) 
    million_cities.append(j[0]) #добавляем содержимое к списку городов
million_cities = tuple(million_cities[1:]) #в первом ряду кусок от названия, отсекаем его
million_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [34]:
# текст запроса
query_5_7 = f'''
WITH temp AS(
                     select a.name,
                     COUNT (*) cnt
                FROM 
                     vacancies v
                LEFT JOIN 
                     areas a ON a.id = v.area_id
                LEFT JOIN 
                     employers e ON e.id = v.employer_id
                WHERE e.name = 'Яндекс'
                      AND a.name in {million_cities}
                GROUP BY 1
                ORDER BY 2
               )
                SELECT *  FROM temp
                UNION ALL
                SELECT 
                     'Total', 
                     sum(cnt)  
                FROM temp
'''

In [35]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

1) Регионы не упорядочены: в значениях можно найти как целые страны (Россия, Казахстан, Беларусь), отдельные субъекты федерации (Московская область, Краснодарский край), и отдельные населённые пункты. 
2) У большого количества работодателей (>8000) не указана сфера деятельности, что тоже негативно сказывается на качестве данных, так как какие-то из сфер деятельности могут в принципе не отслеживаться и не браться в расчёт
3) Большое количество компаний, связанных с разработкой ПО (> 3500). Фактическое их количество может быть больше, но проверить это мы не можем исходя из п. 2

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''
select count (v.id)
from vacancies v
where v.name like '%Данн%' or v.name like '%данн%' or v.name like '%ДАНН%' or v.name like '%DATA%' or v.name like '%Data%' or v.name like '%data%'
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_22052\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [16]:
# текст запроса
query_6_2 = f'''
select
    count(v.name)

from vacancies v

WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

    AND (v.name ILIKE '%junior%'
    OR v.experience ILIKE '%Нет опыта%'
    OR v.employment ILIKE '%Стажировка%')
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [24]:
# текст запроса
query_6_3 = f'''
select count(v.id)
from vacancies v
where
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
)
and  (v.key_skills like '%SQL%'
or v.key_skills like '%postgres%') 
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [30]:
# текст запроса
query_6_4 = f'''
select count(v.id)
from vacancies v
where
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
)
and  v.key_skills ilike '%Python%'
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [36]:
# текст запроса
query_6_5= f'''
select
round(avg(length(key_skills) - length(replace(key_skills,CHR(9),''))+1),2)

from vacancies v

WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
'''

In [37]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [40]:
# текст запроса
query_6_6= f'''
select

  v.experience, ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)

from vacancies v

WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
group by 1
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

***

## Выводы по предметному анализу
1) В БД имеете 1771 вакансия, так или иначе связанные с данными, из них 536 относятся к data science. Из чего делаем вывод, что примерно 30% из всех вакансий связанных с данными так или иначе связаны с data science
2) Из 536 вакансий для дата саентиство 51 доступна для соискателей уровня junior, то есть меньше 10% всех ваканский
3) Исходя из данных выше Python более востребован для работодателей по сравниию с SQL / Postgres (357 против 229 вакансий соответственно). При этом эти выводы нужно принимать со скептицизмом, так как сложно поверить, например в то, что в 200 вакансиях не нужен Python, а SQL нужен меньше, чем в половине вакансий. 
4) С ростом опыта ЗП дата саентиста повышается приблизительно на 70 000 - 100 000 руб. каждые 2-3 отработанных года

# Общий вывод по проекту

In [50]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### 1. Количество ожидаемых навыков соискателя и опыт работы в data science
Посмотрим какое среднее количество навыков ожидает работодатель от соискателя в зависимости от опыта работы (среди тех вакансий, где ключевые навыки is not null)

In [60]:
extra_query_1 = f'''
select v.experience, round(avg(length(key_skills) - length(replace(key_skills,CHR(9),''))+1),2)
from vacancies v
where (
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
group by 1
order by 2
'''

In [61]:
# результат запроса
df = pd.read_sql_query(extra_query_1, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\1894812648.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_1, connection)


experience  round
0         Более 6 лет   5.94
1       От 3 до 6 лет   5.99
2  От 1 года до 3 лет   6.73
3           Нет опыта   7.42

Тут получается немного парадоксальная ситуация - чем больше опыт работы дата саентиста, тем меньше указано ожидаемых ключевых навыков. В среднем от соискателей без опыта работы требуется ~ на 1.5 навыка больше, чем от соискателей с опытом работы более 6 лет.
Моё предполагаемое объяснение такому феномену - в описаниях вакансий для новичков указывается большее количетсво "софт скиллов" - ответсвенность, умение работать в команде и пр., тогда как в вакансиях к опытным саентистам такие требования указываеются реже, потому что им не нужно "доказывать" своё владение софт скиллами.

### 2. Английский язык и data science
Попробуем выяснить, насколько важен английский язык для дата саентиста.

In [69]:
## посмотрим, как он в принципе формулируется в нашем дата сете
extra_query_2 = f'''
select v.key_skills
from vacancies v 
'''

In [70]:
# результат запроса
df = pd.read_sql_query(extra_query_2, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\4065397836.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_2, connection)


key_skills
0      Пользователь ПК\tРабота в команде\tРемонт ноут...
1      Средства криптографической защиты информации\t...
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit
4                                                   None
...                                                  ...
49192  Электронный документооборот\t1С: Предприятие 8...
49193  Оптимизация бизнес-процессов\tСтратегическое м...
49194  Организаторские навыки\tУправление персоналом\...
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...
49196                                               None

[49197 rows x 1 columns]

В нашем случае это "Английский язык"

In [93]:
# Высчитаем долю вакансий, где английский указан в списке навыков в разрезе опыта работы кандидатов
extra_query_3 = f'''
select v.experience, count(v.key_skills ILIKE '%Английский язык%') english_needed, count(v.id) total_vacancies
from vacancies v 
where (
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
    )
group by 1
'''

In [94]:
# результат запроса
df = pd.read_sql_query(extra_query_3, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\3709330608.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_3, connection)


experience  english_needed  total_vacancies
0         Более 6 лет              33               34
1           Нет опыта              26               28
2  От 1 года до 3 лет             194              220
3       От 3 до 6 лет             178              195

Как видно из таблицы выже владение английским - очень важный навык для дата саентиста любого уровня опыта. Отсюда не следует насолько хорош должен быть уровень английского соискателя, поэтому предположим, что оно должно быть достаточным для чтения документации и ведения деловой переписки.

### Пропуски
Высчитаем количество пропусков в полях таблицы vacancies, где они встречаются(key_skills, salary_from, salary_to) при условии, что общая длина таблицы 49197

In [160]:
extra_query_4 = f'''
select count(v.key_skills is null) "key_skills is null"
from vacancies v
where v.key_skills is null
'''

In [161]:
# результат запроса
df = pd.read_sql_query(extra_query_4, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\2771657278.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_4, connection)


key_skills is null
0                8456

Итого, приблизительно 17% пропусков в поле key_skills

In [168]:
extra_query_5 = f'''
select count(v.salary_from is null) "salary_from is null"
from vacancies v
where v.salary_from is null
'''

In [169]:
# результат запроса
df = pd.read_sql_query(extra_query_5, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\4021727357.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_5, connection)


salary_from is null
0                27712

В поле salary_from процент пропусков - 56% что уже много. Такой столбец можно в принципе удалить (в соответсвии с модулем python по очистке данных), он не особо информативен. Мы также рассматриваем только пропуска null, без скрытых.

In [170]:
extra_query_6 = f'''
select count(v.salary_to is null) "salary_to is null"
from vacancies v
where v.salary_to is null
'''

In [171]:
# результат запроса
df = pd.read_sql_query(extra_query_6, connection)
df

C:\Users\vladi\AppData\Local\Temp\ipykernel_10732\3341043337.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(extra_query_6, connection)


salary_to is null
0              36507

В поле salary_to практически 75% всех значений это пропуски.

### Финальный комментарий
2) БД даёт неплохое преставление о географии вакансий (было бы ещё лучше, если таблица areas содержала бы в себе больше столбцов (город, страна, регион, и т.д.)), тогда было бы с большей уверенностью делать выводы о географии поиска работы. Пока что у нас много регионов, где один включает в себя другой (писал об этом ранее)
2) У нас есть надёжная информация о том, какова пропорция требуемых соискателей с разным уровнем работы, что всегда полезно знать. Процент пропусков в поле key_skills также невысок, поэтому на эти данные мы также можем положиться
3) При этом оценивать ЗП по такому датасету сложно исходя из п.1 и также огромного количества пропусков в полях о ЗП. Заполнить такие пропуски будет ошибкой и скорее всего только исказит данные и соотвественно выводы, которые будут по ним сделаны
4) Мы также можем сделать вывод о том, в каих индустриях требуется дата саентисты - эти данные тоже будем считать надёжными, пропусков в них нет